### ISSUE TITLE
COLORS: Lost Memories: loop when displaying 2nd page in custom textbox

### ISSUE THREAD

**Issue by carstene1ns** [EasyRPG/Player#1247]
#### Name of the game:
COLORS: Lost Memories ([download](https://rpgmaker.net/games/9973/))

#### Attach files (as a .zip archive or link them)
Savegame: [Save01.zip](https://github.com/EasyRPG/Player/files/1220620/Save01.zip)

#### Describe the issue in detail and how to reproduce it:
Go right, read the sign (or go right, up the tower, collect the red stone and go back). In both cases a Picture containing a message is shown, press the Action key to advance to second page, then get stuck.
There is running an endless loop, you can hear the message box sound repeating.

![screenshot_0](https://user-images.githubusercontent.com/4691314/29251900-f6483ca4-805d-11e7-8bec-cee2f76e8241.png) → ![screenshot_1](https://user-images.githubusercontent.com/4691314/29251899-f6458202-805d-11e7-9bb3-06db4e1c6efe.png)

This is made of the sign event with 2 pages (39) and 2 common events (8 and 9). So could be a timing problem. Common Event 8 waits for key input and sets some Variables and Switches, Common Event 9 could be ignored, as it only displays and hides the indicator arrow. The Map Event 39 is triggered on Action button, enables the Common Event 8 and on page 2 is set to Autorun on Switch (set by Common Event 8).

**Comment by Ghabry** [comment]
The code after the Key Input "Check, ENTER" in CE 8 is strange because it waits for ENTER but afterwards the code checks if enter was pressed (thats the only possible case) and afterwards if enter was not pressed (impossible)

**Comment by Ghabry** [comment]
Probably the relevant event is Auto-CE 40 which is activated through Switch 20 by Auto-Event 39 Page 2.

CE 40 is responsible for disabling autorun event 39P2 after ENTER is pressed. So my guess is that the common event never gets a chance to run due to a timing problem and E39P2 endlessly repeats instead (as indicated by the endless SE playback)

**Comment by Ghabry** [comment]
The RPG_RT behaviour makes no sense (o'rly). Theres no obvious reason why one Autorun CE can run when an Event CE finishes.
Observed a 2nd behaviour: When the 1st Event CE doesn't yield the interpreter a 2nd Auto-Event CE starts (otherwise not >.<)

**Comment by Ghabry** [comment]
According to my tests the behaviour is (all (common) events are autostart). Adding Parallel processes is left as an exercise ;):
```
E1: Var X+1
E2: Var X+1
E3: Var X+1
```
Result: E1, E2 and E3 run in the same frame. _(expected and implemented)_

```
E1: Var X+1.
E2: Var X+1. Wait 0.0
E3: Var X+1
```
Result: E1 and E2 run in the same frame. E3 never. _(expected and implemented)_

```
E1: Var X+1. Wait 0.0
E2: ...
```
Result: E1 runs. (same as the case above) _(expected and implemented)_

```
C1: Var X+1.
C2: Var X+1.
```
Result: C1 runs and the interpreter hangs (massive lag). C2 never runs(!)

```
C1: Var X+1. Wait 0.0
C2: Var X+1.
```
Result: C1 runs. No lag.

**Now more interesting**
```
E1: Var X+1
E2: Var X+1
C1: Var X+1
C2: Var X+1
```
Result: E1 runs, C1 runs infinitely (massive lag)

```
E1: Var X+1
E2: Var X+1
C1: Var X+1. Wait 0.0
C2: Var X+1
```
Result: E1 runs, C1 runs. Next frame: repeat.

First observation: Only the first auto start common event can ever run, all other CEs never run. _(not implemented, easy fix)_

Second observation: When you have both Autostart Map and Common Events only the first map event runs and the first common event. _(not implemented, tricky. This is what the game uses)_

So having one autorun Common event messes up the whole execution logic

**Comment by Ghabry** [comment]
Thank for these information! I had no time yet to think about your text but my test cases were created with RPG2k3 1.08 because I analyzed this with the DynRPG Event Tracer plugin.

**Comment by fmatthew5876** [comment]
@CherryDT 

I had a few clarifying questions about interpreter behavior

1. In what order does RPG_RT process map events? Currently in player we do player, events, vehicles in that order. Is that right?

2. Does it check autostart conditions for all events before running the update routes? Or does it do this as part of the update route?

3. When is the `trigger=collision` check done? Before or after the event processes a frame of movement?

4. I'm assuming what we call "processed" flag is the "already acted" flag mentioned by you above.

What happens when event A wants to move but event B is in the way but B.processed = false? Does it go an run B's update routine first and then return? Does this also imply that B's event activation checks would happen at this time too?

**Comment by CherryDT** [comment]
I'm not sure I understood 100% the context of each of the questions (because there are for example several ways of "updating", several ways of "checking collision", and so on - RM has often 2 or 3 functions doing almost the same thing with small differences, called in different places), but I think the answers are as follows:

1. Normally: Common events, map events, player, vehicles. Initially after changing map or when exiting game menu: Common events, map events.
2. As part of the update route as described above.
3. After movement. Note that any event simply attempting to move onto another will trigger a part of the other's processing early, out of order, before the passability calculation returns a result, so that a group of events can move together regardless of their ID order without "bumping into each other".
4. Yes. (I checked the chunk ID now, it matches.)

**Comment by fmatthew5876** [comment]
@CherryDT 

It looks like there is a frame delay when an event changes page to when it starts executing.

Lets say EV01 enables switch 1, which is the condition for Page 2 of EV02. Page 2 of EV02 is autostart. It looks like EV02 will wait 1 frame after the switch gets turned on before the autostart event actually gets flagged as "waiting to execute"

When do we actually check page conditions and change the active page? Is it during the event's update routine?

Is the map event's update routine something like this?
```
If (otherPageEnabledByNewConditions()) {
  ChangePage();
  return; //<-skips the frame
}

SetWaitingFlagIfAutostart();
If (AnyEventWaitingOrInterpreterHasActiveEvent()) {
  return; //<-no movement if interp running or will run
}

UpdateMovement();

CheckForCollisionWithPlayer();
```

**Comment by fmatthew5876** [comment]
This is fixed by #1628
